File: app/src/main/java/feelings/guide/ui/RecyclerViewCursorAdapter.java
package feelings.guide.ui;

import android.database.Cursor;
import android.database.DataSetObserver;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import static android.provider.BaseColumns._ID;

public abstract class RecyclerViewCursorAdapter<VH extends RecyclerView.ViewHolder> extends RecyclerView.Adapter<VH> {

    protected final BaseActivity activity;
    private Cursor cursor;
    private boolean isDataValid;
    private int rowIdColumn;

    protected RecyclerViewCursorAdapter(BaseActivity activity) {
        this.activity = activity;
        setHasStableIds(true);
        swapCursor(null);
    }

    protected abstract void onBindViewHolder(VH holder, Cursor cursor);

    @Override
    public void onBindViewHolder(@NonNull VH holder, int position) {
        if (!isDataValid) {
            throw new IllegalStateException("this should only be called when the cursor is valid");
        }
        if (!cursor.moveToPosition(position)) {
            throw new IllegalStateException("couldn't move cursor to position " + position);
        }
        onBindViewHolder(holder, cursor);
    }

    @Override
    public long getItemId(int position) {
        if (isDataValid && cursor != null && cursor.moveToPosition(position)) {
            return cursor.getLong(rowIdColumn);
        }
        return RecyclerView.NO_ID;
    }

    @Override
    public int getItemCount() {
        if (isDataValid && cursor != null) {
            return cursor.getCount();
        } else {
            return 0;
        }
    }

    protected Cursor getCursor() {
        return cursor;
    }

    public void changeCursor(Cursor cursor) {
        Cursor old = swapCursor(cursor);
        if (old != null) {
            old.close();
        }
    }

    protected Cursor swapCursor(Cursor newCursor) {
        if (newCursor == cursor) {
            return null;
        }
        Cursor oldCursor = cursor;
        if (oldCursor != null) {
            oldCursor.unregisterDataSetObserver(dataSetObserver);
        }
        cursor = newCursor;
        if (newCursor != null) {
            newCursor.registerDataSetObserver(dataSetObserver);
            rowIdColumn = newCursor.getColumnIndexOrThrow(_ID);
            isDataValid = true;
            notifyDataSetChanged();
        } else {
            rowIdColumn = -1;
            isDataValid = false;
            notifyDataSetChanged();
        }
        return oldCursor;
    }


    private final DataSetObserver dataSetObserver = new DataSetObserver() {
        @Override
        public void onChanged() {
            isDataValid = true;
            notifyDataSetChanged();
        }

        @Override
        public void onInvalidated() {
            isDataValid = false;
            notifyDataSetChanged();
        }
    };

    private boolean isEmpty() {
        return cursor.getCount() == 0;
    }

    public boolean isNotEmpty() {
        return !isEmpty();
    }
}

File: app/src/main/kotlin/feelings/guide/ui/RecyclerViewCursorAdapter.kt
package feelings.guide.ui

import android.content.Context
import android.database.Cursor
import android.database.DataSetObserver
import android.provider.BaseColumns._ID
import androidx.recyclerview.widget.RecyclerView

abstract class RecyclerViewCursorAdapter<VH : RecyclerView.ViewHolder>
protected constructor(protected val context: Context) :
    RecyclerView.Adapter<VH>() {
    protected var cursor: Cursor? = null
    private var isDataValid: Boolean = false
    private var rowIdColumn: Int = 0


    private val dataSetObserver = object : DataSetObserver() {
        override fun onChanged() {
            isDataValid = true
            notifyDataSetChanged()
        }

        override fun onInvalidated() {
            isDataValid = false
            notifyDataSetChanged()
        }
    }

    private val isEmpty: Boolean
        get() = cursor?.count ?: 0 == 0

    val isNotEmpty: Boolean
        get() = !isEmpty

    protected abstract fun onBindViewHolder(holder: VH, cursor: Cursor)

    override fun onBindViewHolder(holder: VH, position: Int) {
        if (!isDataValid) {
            throw IllegalStateException("this should only be called when the cursor is valid")
        }
        if (!cursor!!.moveToPosition(position)) {
            throw IllegalStateException("couldn't move cursor to position $position")
        }
        onBindViewHolder(holder, cursor!!)
    }

    override fun getItemId(position: Int): Long {
        return if (isDataValid && cursor!!.moveToPosition(position))
            cursor!!.getLong(rowIdColumn)
        else RecyclerView.NO_ID
    }

    override fun getItemCount(): Int {
        return if (isDataValid) cursor?.count ?: 0 else 0
    }

    protected fun swapCursor(newCursor: Cursor): Cursor? {
        if (newCursor === cursor) {
            return null
        }

        val oldCursor = cursor
        oldCursor?.unregisterDataSetObserver(dataSetObserver)

        cursor = newCursor
        newCursor.registerDataSetObserver(dataSetObserver)
        rowIdColumn = newCursor.getColumnIndexOrThrow(_ID)
        isDataValid = true

        notifyDataSetChanged()
        return oldCursor
    }

    fun close() {
        cursor?.unregisterDataSetObserver(dataSetObserver)
        cursor?.close()
        notifyDataSetChanged()
    }
}